### PR TITLE
cmake: add a memcheck option for ctest

### PR DESCRIPTION
**Pull request by NitzanMordhai**
Adding the option to run the ctest with valgring so we will be able to find un-initialized variables.
To run the unit-test with memory check:
ctest -R name_test -T memcheck -j 80

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -733,6 +733,13 @@ add_subdirectory(erasure-code)
 
 # Support/Tools
 if(WITH_TESTS)
+  find_program(CMAKE_MEMORYCHECK_COMMAND valgrind)
+  message(STATUS "Executing test suite with Valgrind (${CMAKE_MEMORYCHECK_COMMAND})")
+  set(MEMORYCHECK_COMMAND_OPTIONS "--error-exitcode=1 --leak-check=full --suppressions=${CMAKE_SOURCE_DIR}/qa/valgrind.supp")
```
Comment (rzarzynski) [inline src/CMakeLists.txt:738]: nit: not sure we need to repeat the `--suppression` here as `MEMORYCHECK_SUPPRESSION_FILE` is defined below.

**Review by rzarzynski — COMMENTED** [review]
I very, very like this idea! I hope that ultimately every make check run will happen under valgrind.

I understand there will be a transition period in which we cannot enable (merge) the memcheck because of existing, hidden bugs. This is the main reason behind not approving this immediately.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -733,6 +733,15 @@ add_subdirectory(erasure-code)
 
 # Support/Tools
 if(WITH_TESTS)
+  find_program(CMAKE_MEMORYCHECK_COMMAND valgrind)
+  message(STATUS "Executing test suite with Valgrind (${CMAKE_MEMORYCHECK_COMMAND})")
+  set(MEMORYCHECK_COMMAND_OPTIONS
+  "--trace-children=yes --leak-check=full --soname-synonyms=somalloc=\*tcmalloc\* \
```
Comment (rzarzynski) [inline src/CMakeLists.txt:739]: Hmm, is overriding `--trace-children` (first `yes`, then `no`) intentional?

Comment (NitzanMordhai) [reply src/CMakeLists.txt]: no, fixing

File: CMakeLists.txt [hunk 3]
```diff
@@ -753,6 +753,14 @@ endif(WITH_SELINUX)
 add_custom_target(tests
   COMMENT "Building tests")
 enable_testing()
+include(CTest)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply CMakeLists.txt:756]: also, we might want to wire `BUILD_TESTING` with `WITH_TESTS`.

**Review by tchaikov — COMMENTED** [review]
left a few comments.

File: CMakeLists.txt [hunk 4]
```diff
@@ -753,6 +753,14 @@ endif(WITH_SELINUX)
...
+include(Dart)
+set(CMAKE_CTEST_COMMAND ctest)
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply CMakeLists.txt:758]: @tchaikov i thought we will want to run the memcheck only when "-T memcheck" is pass

[the remark this answers is not in the document]
Comment (tchaikov) [reply CMakeLists.txt:758]: i meant to refer the line of `set(CMAKE_CTEST_COMMAND ctest)`.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -733,6 +733,15 @@ add_subdirectory(erasure-code)
...
 if(WITH_TESTS)
+  find_program(CMAKE_MEMORYCHECK_COMMAND valgrind)
+  message(STATUS "Executing test suite with Valgrind (${CMAKE_MEMORYCHECK_COMMAND})")
+  set(MEMORYCHECK_COMMAND_OPTIONS
+  "--trace-children=no --leak-check=full --soname-synonyms=somalloc=*tcmalloc* \
+  --child-silent-after-fork=yes")
+  set(MEMORYCHECK_SUPPRESSIONS_FILE "${CMAKE_SOURCE_DIR}/qa/valgrind.supp")
+  set(memcheck_command "${CMAKE_MEMORYCHECK_COMMAND} ${CMAKE_MEMORYCHECK_COMMAND_OPTIONS}")
+  include(Dart)
+  separate_arguments(memcheck_command)
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/CMakeLists.txt:744]: Do we need it in src/CMakeList.txt as well, or I can remove it?

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply CMakeLists.txt:756]: please refer to https://cmake.org/cmake/help/latest/module/CTest.html

File: src/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/CMakeLists.txt:744]: After some read from the link you provided about BUILD_TEST, I can remove that re-configure of valgrind here and leave the main CMakeList with it.

**Comment by tchaikov** [comment]
> I very, very like this idea! I hope that ultimately every make check run will happen under valgrind.
> 
> I understand there will be a transition period in which we cannot enable (merge) the memcheck because of existing, hidden bugs. This is the main reason behind not approving this immediately.

i would prefer running the tests with ASan. and run them with valgrind with a daily cron job. but yeah, if we have the ~~horse~~ machine power to run them for every single PR with valgrind, why not?

File: CMakeLists.txt [hunk 6]
```diff
@@ -753,6 +753,14 @@ endif(WITH_SELINUX)
 add_custom_target(tests
   COMMENT "Building tests")
 enable_testing()
+include(CTest)
+include(Dart)
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply CMakeLists.txt]: removed

File: CMakeLists.txt [hunk 7]
```diff
@@ -753,6 +753,14 @@ endif(WITH_SELINUX)
...
+set(CMAKE_CTEST_COMMAND ctest)
+find_program(MEMORYCHECK_COMMAND valgrind)
+set(MEMORYCHECK_COMMAND_OPTIONS
+  "--trace-children=no --leak-check=full --soname-synonyms=somalloc=*tcmalloc* \
+  --child-silent-after-fork=yes")
+set(MEMORYCHECK_SUPPRESSIONS_FILE "${CMAKE_SOURCE_DIR}/qa/valgrind.supp")
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply CMakeLists.txt]: done

**Comment by cbodley** [comment]
> i would prefer just running the tests with ASan, and run them with valgrind with a daily cron job. but yeah, if we have the ~horse~ machine power to run them for every single PR with valgrind, why not?

i'd love to see asan/ubsan enabled for make check. i've always had trouble getting ceph to start up cleanly with asan enabled though, so i imagine this would be a large project

running under valgrind is a lot slower, especially during startup, so i'd like to see how much longer the checks take when it's enabled. it's important that we have timely feedback on PRs, so i wouldn't want valgrind to add hours to the runtime

File: cmake/modules/CTestScript.cmake [hunk 8]
```diff
@@ -0,0 +1,8 @@
+function(build_tests)
```
Comment (tchaikov) [inline cmake/modules/CTestScript.cmake:1]: sorry for the confusion. i think we can either put the body of this function into this file as it is. and let user to call it like
```console
ctest -S cmake/CTestScript.cmake ...
```
or just include this file at the top-level `CMakeLists.txt`. again, i still think it might be better to keep `include(CTest)` in this file.

File: src/tools/CMakeLists.txt [hunk 9]
```diff
@@ -24,7 +24,9 @@ if(NOT WIN32)
       neorados.cc)
   add_executable(neorados ${neorados_srcs})
   target_link_libraries(neorados libneorados ${FMT_LIB} ${CMAKE_DL_LIBS})
-  add_dependencies(tests neorados)
+  if(WITH_TESTS)
+    add_dependencies(tests neorados)
+  endif()
```
Comment (tchaikov) [inline src/tools/CMakeLists.txt:29]: this change can be submitted in another PR.

File: CMakeLists.txt [hunk 10]
```diff
@@ -749,15 +749,19 @@ if(WITH_SELINUX)
   add_subdirectory(selinux)
 endif(WITH_SELINUX)
 
+if(WITH_TESTS)
```
Comment (tchaikov) [inline CMakeLists.txt:752]: this change does not belong to this PR.

**Comment by tchaikov** [comment]
> > i would prefer just running the tests with ASan, and run them with valgrind with a daily cron job. but yeah, if we have the ~horse~ machine power to run them for every single PR with valgrind, why not?
> 
> i'd love to see asan/ubsan enabled for make check. i've always had trouble getting ceph to start up cleanly with asan enabled though, so i imagine this would be a large project

i am surprised to learn this actually. does this also apply to our unit tests? or just the servers ? AFAICT, crimson should already be using ASan, as seastar uses ASan when it is built in Debug mode.

> 
> running under valgrind is a lot slower, especially during startup, so i'd like to see how much longer the checks take when it's enabled. it's important that we have timely feedback on PRs, so i wouldn't want valgrind to add hours to the runtime

yup, i share the same concern with you.

File: cmake/modules/CTestScript.cmake [hunk 11]
```diff
@@ -6,3 +6,9 @@ function(build_tests)
   set(MEMORYCHECK_SUPPRESSIONS_FILE  
     "${CMAKE_SOURCE_DIR}/qa/valgrind.supp")
 endfunction()
+
+function(build_test_asan)
```
Comment (tchaikov) [inline cmake/modules/CTestScript.cmake:10]: not sure how far you get, but i ran into a link failure in #56241
